docs/book/v3/cookbook/double-pass-middleware.md

```diff
@@ -0,0 +1,181 @@
+
+> ### What is Double-Pass Middleware?
+>
+> Double pass middleware receives both the request and a response in addition to
+> the handler, and passes both the request and response to the handler when
+> invoking it:
+>
+> ```php
+> function (ServerRequestInterface $request, ResponseInterface $response, callable $next)
+> {
+>     $response = $next($request, $response);
+>     return $response->withHeader('X-Test', time());
+> }
+> ```
+>
+> It is termed "double pass" because you pass _both_ the request _and_ response when
+> delegating to the next layer.
+
+## doublePassMiddleware function
+
+zend-stratigility v2.2 and v3.0 ship a utility function,
+`Zend\Stratigility\doublePassMiddleware()`, that will decorate a callable
+double-pass middleware using a `Zend\Stratigility\Middleware\DoublePassMiddlewareDecorator` 
+instance; this latter is a PSR-15 impelementation, and can thus be used in your
+middleware pipelines.
+
+The function (and class) also expects a [PSR-7](https://www.php-fig.org/psr/psr-7/)
+`ResponseInterface` instance as a second argument; this is then passed as the
+`$response` argument to the double-pass middleware. The following examples
+demostrate both piping and routing to double pass middleware using this
+technique, and using zend-diactoros to provide the response instance.
+
+```php
+use Zend\Diactoros\Response;
+use function Zend\Stratigility\doublePassMiddleware;
+
+$app->pipe(doublePassMiddleware(function ($request, $response, $next) {
+    $response = $next($request, $response);
+    return $response->withHeader('X-Clacks-Overhead', 'GNU Terry Pratchett');
+}, new Response())); // <-- note the response
+
+$app->get('/api/ping', doublePassMiddleware(function ($request, $response, $next) {
+    return new Response\JsonResponse([
+        'ack' => time(),
+    ]);
+}, new Response())); // <-- note the response
+```
+
+## Double-Pass Middleware Services
+
+What if you're piping or routing to a _service_ &mdash; for instance, a class
+provided by a third-party implementation?
+
+In this case, you have one of two options:
+
+- Decorate the middleware before returning it from the factory that creates it.
+- Use a [delegator factory](../features/container/delegator-factories.md) to
+  decorate the middleware.
+
+### Decorating via factory
+
+If you have control of the factory that creates the double-pass middleware you
+will be using in your application, you can use the strategy outlined above to
+decorate your middleware before returning it, with one minor change: you can
+pull a response factory from the container as well.
+
+To demonstrate:
+
+```php
+use Psr\Container\ContainerInterface;
+use Psr\Http\Message\ResponseInterface;
+
+use function Zend\Stratigility\doublePassMiddleware;
+
+class SomeDoublePassMiddlewareFactory
+{
+    public function __invoke(ContainerInterface $container)
+    {
+        // Create the middleware instance somehow. This example
+        // assumes it is in `$middleware` when done.
+
+
```

Comment:
Remove one empty line, one should be enough.

Reply:
Fixed.